src/org/traccar/database/NotificationManager.java

```diff
@@ -18,57 +19,70 @@
 import java.lang.reflect.Field;
 import java.lang.reflect.Modifier;
 import java.sql.SQLException;
-import java.util.Collection;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
-import java.util.concurrent.locks.ReadWriteLock;
-import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 import org.traccar.Context;
 import org.traccar.helper.Log;
 import org.traccar.model.Event;
 import org.traccar.model.Notification;
 import org.traccar.model.Position;
+import org.traccar.model.Typed;
 import org.traccar.notification.NotificationMail;
 import org.traccar.notification.NotificationSms;
 
-public class NotificationManager {
-
-    private final DataManager dataManager;
-
-    private final Map<Long, Set<Notification>> userNotifications = new HashMap<>();
-
-    private final ReadWriteLock notificationsLock = new ReentrantReadWriteLock();
+public class NotificationManager extends ExtendedObjectManager<Notification> {
 
     public NotificationManager(DataManager dataManager) {
-        this.dataManager = dataManager;
-        refresh();
+        super(dataManager, Notification.class);
+    }
+
+    private Set<Long> getEffectiveNotifications(long userId, long deviceId) {
+        Set<Long> result = new HashSet<>();
+        Set<Long> deviceNotifications = getAllDeviceItems(deviceId);
+        for (long itemId : getUserItems(userId)) {
+            if (getById(itemId).getAllways() || deviceNotifications.contains(itemId)) {
+                result.add(itemId);
+            }
+        }
+        return result;
     }
 
     public void updateEvent(Event event, Position position) {
         try {
-            dataManager.addObject(event);
+            getDataManager().addObject(event);
         } catch (SQLException error) {
             Log.warning(error);
         }
 
-        Set<Long> users = Context.getPermissionsManager().getDeviceUsers(event.getDeviceId());
+        long deviceId = event.getDeviceId();
+        Set<Long> users = Context.getPermissionsManager().getDeviceUsers(deviceId);
         for (long userId : users) {
             if (event.getGeofenceId() == 0 || Context.getGeofenceManager() != null
                     && Context.getGeofenceManager().checkItemPermission(userId, event.getGeofenceId())) {
-                Notification notification = getUserNotificationByType(userId, event.getType());
-                if (notification != null) {
-                    if (notification.getWeb()) {
-                        Context.getConnectionManager().updateEvent(userId, event);
-                    }
-                    if (notification.getMail()) {
-                        NotificationMail.sendMailAsync(userId, event, position);
+                boolean webSent = false;
```

Comment:
This flags help to not do extra iterations and not duplicate notifications

Reply:
One small comment here. I think if we move "sent" to be a prefix, it will be easier to read the code.